tuf/client_rework/metadata_bundle.py

```diff
@@ -0,0 +1,450 @@
+# Copyright the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""TUF client bundle-of-metadata
+
+MetadataBundle keeps track of current valid set of metadata for the client,
+and handles almost every step of the "Detailed client workflow" in the TUF
+specification (the remaining steps are download related). The bundle takes
+care of persisting valid metadata on disk, loading local metadata from disk
+and deleting invalid local metadata.
+
+Loaded metadata can be accessed via the index access with rolename as key
+or, in the case of top-level metadata using the helper properties like
+'MetadataBundle.root'
+
+Metadata can be loaded into bundle by two means:
+ * loading from local storage: load_local_metadata()
+   (and, in the case of root metadata, the constuctor)
+ * updating from remote repository: update_metadata()
+
+The rules for top-level metadata are
+ * Metadata is loadable only if metadata it depends on is loaded
+ * Metadata is immutable if any metadata depending on it has been loaded
+ * Loading from local storage must be attempted before updating from remote
+ * Updating from remote is never required
+
+Exceptions are raised if metadata fails to load in any way (except in the
+case of local loads -- see load_local_metadata()).
+
+Example (with hypothetical download function):
+
+>>> # Load local root
+>>> bundle = MetadataBundle("path/to/metadata")
+>>>
+>>> # load more root versions from remote
+>>> with download("root", bundle.root.signed.version + 1) as f:
+>>>     bundle.update_metadata(f.read())
+>>> with download("root", bundle.root.signed.version + 1) as f:
+>>>     bundle.update_metadata(f.read())
+>>>
+>>> # Finally, no more roots from remote
+>>> bundle.root_update_finished()
+>>>
+>>> # load local timestamp, then update it
+>>> bundle.load_local_metadata("timestamp")
+>>> with download("timestamp") as f:
+>>>     bundle.update_metadata(f.read())
+>>>
+>>> # load local snapshot, then update it if needed
+>>> if not bundle.load_local_metadata("snapshot"):
+>>>     # load snapshot (consistent snapshot not shown)
+>>>     with download("snapshot") as f:
+>>>         bundle.update_metadata(f.read())
+>>>
+>>> # load local targets, then update it if needed
+>>> if not bundle.load_local_metadata("targets"):
+>>>     version = bundle.snapshot.signed.meta["targets.json"]["version"]
+>>>     with download("snapshot", version + 1) as f:
+>>>         bundle.update_metadata(f.read())
+>>>
+>>> # Top level metadata is now fully loaded and verified
+
+
+TODO:
+ * Delegated targets are implemented but they are not covered
+   by same immutability guarantees: the top-level metadata is handled
+   by hard-coded rules (can't update root if snapshot is loaded)
+   but delegations would require storing the delegation tree ...
+ * exceptions are all over the place and not thought out at all
+ * usefulness of root_update_finished() can be debated: it could be done
```

Comment:
Yea, I agree and I don't see where are you using it now?

Reply:
Updater calls it, it is a public method.
 If we want to be strict with the spec and do what it calls "protection from fast forward attacks" , I think this is the place. Although I admit, you already got me here because I forgot to call `root_update_finished()` and got an error :)

Reply:
it's a public method that client should currently call after final root has been loaded -- according to spec we verify root expiry at this point... Practically we could do it when the next metadata (timestamp) is about to be loaded but it looks a bit goofy: caller is trying to load timestamp but gets an error saying root is no longer valid.